### PR TITLE
add ability to import ics attachments into the calendar

### PR DESCRIPTION
**Pull request by ChristophWurst**
fixes: https://github.com/owncloud/mail/issues/79

TODO:
- [x] show import button
- [x] download attachment file
- [x] select calendar
- [x] push ics content to the dav endpoint
- [x] show import button only if caldav is enabled/available
- [x] add ES6 promise poly fill
- [x] list only writable calendars

BUGS:
- [x] fix attachment text width if the import button is shown
- [x] popover position
- [x] some .ics file imports fail ``HTTP/1.1 400 Every VEVENT in this object must have identical UIDs``

FUTURE IDEAS
- Show a preview of the calendar data

### CODE REVIEW

**Comment by ChristophWurst** [comment]
bildschirmfoto von 2016-05-04 11-32-26
bildschirmfoto von 2016-05-04 11-30-27
bildschirmfoto von 2016-05-04 11-30-19

File: js/service/davservice.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		xmlNamespaces: {
+			'DAV:': 'd',
+			'urn:ietf:params:xml:ns:caldav': 'c',
+			'http://apple.com/ns/ical/': 'aapl',
+			'http://owncloud.org/ns': 'oc',
+			'http://calendarserver.org/ns/': 'cs'
+		}
+	});
+	var props = [
+		'{' + dav.NS_DAV + '}displayname',
```
Comment (DeepDiver1975) [inline js/service/davservice.js]: @ChristophWurst I assume dav.NS_* are all undefined

Comment (ChristophWurst) [inline js/service/davservice.js]: Those are only defined in calendar: https://github.com/owncloud/calendar/blob/8e32fee4f2bf1346e987c23d2d0ab05cebcefd40/js/app/service/davclient.js#L39-L43

**Comment by ChristophWurst** [comment]
bildschirmfoto von 2016-05-04 15-52-17

**Comment by jancborchardt** [comment]
Ooooh that's awesome! :)

I would now even (maybe) for the ics files consolidate the download and save to Files action into a 3-dot menu cause they are leaa important now.

**Comment by ChristophWurst** [comment]
> I would now even (maybe) for the ics files consolidate the download and save to Files action into a 3-dot menu cause they are leaa important now.

That's my plan, but I'd rather do that in a follow up PR

**Comment by ChristophWurst** [comment]
@DeepDiver1975 when PUTing the file to the selected calendar, I get the following error:
```
{"reqId":"dnA7v9MDCVi6pPDB+eeu","remoteAddr":"::1","app":"webdav","message":"Exception: {\"Message\":\"HTTP\\\/1.1 400 Namespace prefix c on calendar-description is not defined\\n on line 5, column 28\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\BadRequest\",\"Code\":0,\"Trace\":\"#0 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpPropFind(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#1 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#2 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(459): Sabre\\\\Event\\\\EventEmitter->emit('method:PROPFIND', Array)\\n#3 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#4 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/server.php(162): Sabre\\\\DAV\\\\Server->exec()\\n#5 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(29): OCA\\\\DAV\\\\Server->exec()\\n#6 \\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/remote.php(164): require_once('\\\/home\\\/christoph...')\\n#7 {main}\",\"File\":\"\\\/home\\\/christoph\\\/workspace\\\/owncloud\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php\",\"Line\":322,\"User\":\"user\"}","level":4,"time":"2016-05-04T10:10:09+00:00","method":"PROPFIND","url":"\/remote.php\/dav\/calendars\/user","user":"user"}
```
Any idea what I'm doing wrong?

**Comment by ChristophWurst** [comment]
I'd like to refactor the code a little bit to allow registering different attachment preview generators so we can preview other file types too in the future. To keep this PR reviewable, I will do that in a follow-up PR.

**Comment by ChristophWurst** [comment]
- [ ] TODO: update changelog

**Comment by ChristophWurst** [comment]
@jyaworski @jancborchardt @Gomez @tahaalibra @DeepDiver1975 this is ready to review :-)

File: js/require_config.js [hunk 2]
```diff
@@ -21,11 +21,20 @@
...
+			ical: 'vendor/ical.js/build/ical.min',
 			marionette: 'vendor/backbone.marionette/lib/backbone.marionette',
 			underscore: 'vendor/underscore/underscore',
 			text: 'vendor/text/text'
+			},
+			davclient: {
+				exports: 'dav'
+			},
+			ical: {
+				exports: 'ICAL'
```
Comment (ChristophWurst) [inline js/require_config.js]: - [ ] rebase destroyed the json structure

**Comment by tahaalibra** [comment]
selection_002
it is working great here, it shows the + button

selection_003
but here it doesn't show + button for the same file ( here the difference is that the email was a forwarded email, but in the first case where it is showing + button, it was a new email)

**Comment by ChristophWurst** [comment]
@tahaalibra the + Button is only shown for the MIME type 'text/calendar'. Could you please check the MIME type of that file? I assume it's a different type as the calendar icon is also not shown.

**Comment by tahaalibra** [comment]
i am talking about file meetup.ics, its the same file on both email (ignore contact_birt..)

**Comment by ChristophWurst** [comment]
Me too. Check the icon, it's not the same ;-) Could you please debug this so we know what exact MIME type the second file has? Thanks

**Comment by Gomez** [comment]
Works perfect for me. Lets get it in and solve following problems on the way...

**Pull request merged**